Fix 4628

micw opened the pull request:
## Description:


**Related issue (if applicable):** fixes #4628

## Checklist:

  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] Tests have been added to verify that the new code works.

micw committed:
Fix 4628

micw committed:
Add requirements to setup.py

micw commented:
That's a breaking change for non-english component names!

micw committed:
Fix test

andrey-git commented:
Note that unidecode is GPL. There are other libraries like python-slugify

micw commented:
@andrey-git good point, I'll check the other lib

micw committed:
Switch from unidecode to python_slugify for licensing reasons

micw commented:
@andrey-git I changed to python_slugify. Works the same, tests should be ok as well.

I have the problem that it works with hass and with travis but not with tox when I run it locally:
Any ideas?

```
py35 runtests: commands[0] | py.test --timeout=30 --duration=10 --cov --cov-report= tests/util/test_init.py
Traceback (most recent call last):
  File "/home/mwyraz/git/home-assistant/.tox/py35/lib/python3.5/site-packages/_pytest/config.py", line 325, in _getconftestmodules
    return self._path2confmods[path]
KeyError: local('/home/mwyraz/git/home-assistant/tests/util/test_init.py')

During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/mwyraz/git/home-assistant/.tox/py35/lib/python3.5/site-packages/_pytest/config.py", line 325, in _getconftestmodules
    return self._path2confmods[path]
KeyError: local('/home/mwyraz/git/home-assistant/tests/util')

During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/mwyraz/git/home-assistant/.tox/py35/lib/python3.5/site-packages/_pytest/config.py", line 356, in _importconftest
    return self._conftestpath2mod[conftestpath]
KeyError: local('/home/mwyraz/git/home-assistant/tests/conftest.py')

During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/mwyraz/git/home-assistant/.tox/py35/lib/python3.5/site-packages/_pytest/config.py", line 362, in _importconftest
    mod = conftestpath.pyimport()
  File "/home/mwyraz/git/home-assistant/.tox/py35/lib/python3.5/site-packages/py/_path/local.py", line 662, in pyimport
    __import__(modname)
  File "/home/mwyraz/git/home-assistant/.tox/py35/lib/python3.5/site-packages/_pytest/assertion/rewrite.py", line 216, in load_module
    py.builtin.exec_(co, mod.__dict__)
  File "/home/mwyraz/git/home-assistant/tests/conftest.py", line 11, in <module>
    from homeassistant import util, setup
  File "/home/mwyraz/git/home-assistant/homeassistant/util/__init__.py", line 13, in <module>
    from slugify import slugify as python_slugify
ImportError: No module named 'slugify'
ERROR: could not load /home/mwyraz/git/home-assistant/tests/conftest.py

ERROR: InvocationError: '/home/mwyraz/git/home-assistant/.tox/py35/bin/py.test --timeout=30 --duration=10 --cov --cov-report= tests/util/test_init.py'
```

andrey-git commented:
Compare your output with what Travis does: https://travis-ci.org/home-assistant/home-assistant/jobs/217733555

Note that there are broken tests there.

Also, I would get a blessing for this change from @balloob before spending more time on it.

micw committed:
Handling colons in slugify, add test

micw committed:
Fix lint

micw commented:
@andrey-git yes, the test is another thing. There are other tests that expects some slugify behaviour for non-alphanumeric chars. E.g. this particular test expected the colon in the mac to be removed, my slugify replaces it with a dash. I dig into these and add special handling.

Nevertheless I still get the local error and have no glue, why that lib is not found by tox within my venv.

micw committed:
mend

micw committed:
Make new slugify more compatible to old

micw committed:
Make new slugify more compatible to old

micw commented:
@andrey-git got the tests green now. There where many implicit assumptions on the slugify function in many tests. I added all that failed to test_init and changed the new implementation to be compatible with the old one for ASCII component names.

Since Non-ASCII names are handled now, it's still a breaking change.
